src/analysis/souffle/taint_with_auth_logic.dl

```diff
@@ -0,0 +1,114 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+// This file gives rules that define a check that does information flow control,
+// but uses authorization logic for downgrading.
+
+#include "flat_graph_ir.dl"
+#include "auth_logic_interface.dl"
+
+#ifndef TAINT_WITH_AUTH_LOGIC
+#define TAINT_WITH_AUTH_LOGIC
+
+//-----------------------------------------------------------------------------
+// Type Declarations
+//-----------------------------------------------------------------------------
+// A taint Tag, such as "video_data"
+.type Tag <: symbol
+
+// An information flow Label which in this model is a set of taint Tags
+.type Label <: symbol
+
+//-----------------------------------------------------------------------------
+// Predicate Declarations
+//-----------------------------------------------------------------------------
+// There is a claim that `accessPath` definitely has `tag` taint.
+.decl claimHasTag(accessPath: AccessPath, tag: Tag)
```

Comment:
In Authorization Logic, it's supposed to be the case that a claim that a `tag` is present on a particular `accessPath` should be attributed to some `principal`, right? Shouldn't a `principal` be present in the parameters of this relation?

Reply:
See refactor in latest commit. The analysis in `taint_with_auth_logic.dl` now shares many of the relations from `taint.dl` which it includes. It implements a separate version of `mayHaveTag` called `mayHaveTagAL` so that this analysis involving authorization logic can exist at the same time as the regular taint tracking one. 

`mayHaveTagAL` is now implemented in terms of edges. There is now also a separate rule in `edges_from_graph.dl` that populates edges from graphs in the low-level IR defined in `flat_graph_ir.dl`

If you'd rather not keep both of these live and just have the authorization logic one live, then I think it would make sense to just merge `taint_with_auth_logic.dl` into `taint.dl`. Let me know if that's what you'd prefer over this way.

Reply:
I think it makes sense to merge them together. We might need to do some refactoring of taint.dl though.

Reply:
This is done, right?